source/SkiaSharp.Views/SkiaSharp.Views.WPF/SKGLElement.cs

```diff
@@ -0,0 +1,222 @@
+
+		[Category("Appearance")]
+		public event EventHandler<SKPaintGLSurfaceEventArgs> PaintSurface;
+
+		private SKSizeI GetSize()
+		{
+			var currentWidth = ActualWidth;
+			var currentHeight = ActualHeight;
+
+			if (currentWidth < 0 ||
+				currentHeight < 0)
+			{
+				currentWidth = 0;
+				currentHeight = 0;
+			}
+
+			PresentationSource source = PresentationSource.FromVisual(this);
+
+			double dpiX = 1.0;
+			double dpiY = 1.0;
+			if (source != null)
+			{
+				dpiX = source.CompositionTarget.TransformToDevice.M11;
+				dpiY = source.CompositionTarget.TransformToDevice.M22;
+			}
+
+			return new SKSizeI((int)(currentWidth * dpiX), (int)(currentHeight * dpiY));
+		}
+
+		protected override void OnRender(DrawingContext drawingContext)
+		{
+			if (grContext != null)
+			{
+				grContext.ResetContext();
+			}
+			base.OnRender(drawingContext);
+		}
+
+		protected virtual void OnPaint(TimeSpan e)
+		{
+			if (disposed)
+			{
+				return;
+			}
+			if (designMode)
+			{
+				return;
+			}
+
+			// create the contexts if not done already
+			if (grContext == null)
+			{
+				var glInterface = GRGlInterface.Create();
+				grContext = GRContext.CreateGl(glInterface);
+			}
+
+			// get the new surface size
+			var newSize = GetSize();
+
+			GL.ClearColor(Color4.Transparent);
+			GL.Clear(ClearBufferMask.ColorBufferBit);
+			GL.Clear(ClearBufferMask.ColorBufferBit | ClearBufferMask.DepthBufferBit | ClearBufferMask.StencilBufferBit);
+
+			// manage the drawing surface
+			if (renderTarget == null || lastSize != newSize || !renderTarget.IsValid)
+			{
+
+				// create or update the dimensions
+				lastSize = newSize;
+
+				GL.GetInteger(GetPName.FramebufferBinding, out var framebuffer);
+				GL.GetInteger(GetPName.StencilBits, out var stencil);
+				GL.GetInteger(GetPName.Samples, out var samples);
+				var maxSamples = grContext.GetMaxSurfaceSampleCount(colorType);
+				if (samples > maxSamples)
+					samples = maxSamples;
+				glInfo = new GRGlFramebufferInfo((uint)framebuffer, colorType.ToGlSizedFormat());
+
+				// destroy the old surface
+				surface?.Dispose();
+				surface = null;
+				canvas = null;
+
+				// re-create the render target
+				renderTarget?.Dispose();
+				renderTarget = new GRBackendRenderTarget(newSize.Width, newSize.Height, samples, stencil, glInfo);
+			}
+
+			// create the surface
+			if (surface == null)
+			{
+				surface = SKSurface.Create(grContext, renderTarget, surfaceOrigin, colorType);
+				canvas = surface.Canvas;
+			}
+
+			using (new SKAutoCanvasRestore(canvas, true))
+			{
+				// start drawing
+				OnPaintSurface(new SKPaintGLSurfaceEventArgs(surface, renderTarget, surfaceOrigin, colorType));
+			}
+
+			// update the control
+			canvas.Flush();
+		}
+
+		protected virtual void OnPaintSurface(SKPaintGLSurfaceEventArgs e)
+		{
+			// invoke the event
+			PaintSurface?.Invoke(this, e);
+		}
+
+		private bool disposed = false;
+
+
+		protected virtual void Dispose(bool disposing)
```

Comment:
WPF UI worlds don't really use Dispose as a mechanism for doing things. Unload is resetting already? I am not 100% confident of how things work in the worlds of heavy graphics objects.